### ISSUE TITLE
Trans component documentation unclear

### ISSUE THREAD

**Issue by ctowler157** [i18next/react-i18next-gitbook#15]
Following what little context the trans component documentation gives, I cannot get any kind of functionality out of the Trans component other than explicitly rendering its children.  There are no examples for how this is supposed to be used within a component (need more context than just showing what the Trans component JSX would look like); this doesn't even show how the Trans component-specific strings fit into your i18n initialization object, just shows a key and a value without any context. The "working example" does not illustrate how to connect your i18n configuration to the trans component, so it is entirely unclear how it even translates the paragraph (if it even does, I haven't tried cloning it and running it because I'd rather understand it first)

This seems like such a useful tool, but all of the documentation is incredibly hard to follow, and the examples that are linked out to are rarely illustrative. I lost my patience with this specific component, but the general theme of my complaints seem to be present throughout the documentation. To be useful to people who don't already understand the tool, the documentation should contain more context in the code samples. This feels more like I'm trying to solve a jigsaw puzzle than following any instructions.

https://react.i18next.com/components/trans-component

**Comment by jamuhl** [comment]
https://github.com/i18next/react-i18next/issues/458

**Comment by jamuhl** [comment]
Can you elaborate a little more?

https://react.i18next.com/components/trans-component

Shows it is there to translate a combination of text and components -> eg. link inside some text, or bold writing inside text. Basically shows how to replace a regular jsx for translation.

It also shows how to get to the correct string.

Endless samples show how to setup a project: https://github.com/i18next/react-i18next/tree/master/example

might be important to know...react-i18next is "just" a UI framework integration of https://i18next.com -> so all translation specifics or init options are documented there

**Comment by jamuhl** [comment]
what is `react-i18-next` ?!! you mean i18next.use(reactI18nextModule)?!?

**Comment by jamuhl** [comment]
i guess that should explain most: https://react.i18next.com/components/overview

**Comment by jamuhl** [comment]
I really know the documentation should be easier...but really have a hard time improving that.

Hm...an option might be i add a page explaining from using just `i18next.t` what you get by adding the next component, like:

`i18next.t` -> `Trans` -> `translate` -> `i18nextProvider or i18next.use(reactI18nextModule)` guess that might help. What do you think?

**Comment by juanecabellob** [comment]
Yes, I meant `reactI18nextModule` not `react-i18-next`. It explains very well the dependencies and how are they interconnected, but not what's the minimum and whether I should use all of them or just one or two components. 

Let me get back to you on that. I'll integrate it to my project and think bottom-up of an example guide. Also, like always sometimes it's not just adding things on top but maybe rethink the whole documentation itself, as the amount of documents is growing and that forces to split attention into different files without a "reader-flow". Happy to help on this once I successfully integrated the project and understand better the whole framework.

**Comment by jamuhl** [comment]
wow...that really would be awesome and a huge help. You made my day...

**Comment by juanecabellob** [comment]
So I fully understand how the pieces work and I propose a Getting started guide that is actually a tutorial that contains the following:
* The best way to start understanding everything is by just starting with a bottom up tutorial:
     * First just introduce two things: The `t` function and the configuration needed to use it. Given that the `t` function can be get with just the `i18next` it would be a very good starting point to raise awareness of the importance to understand the basic concepts on which `i18next` is based and work just with that library without `react-i18next`. I propose this as too many concepts and tools come from `i18next` and without having read that it's much harder to understand the rest of the components of `react-i18next `.  A simple example to start with would be to have a form of which translations for its labels are required and placeholders are required.
    * Now it's time to introduce `react-i18next`. Stating that it doesn't feel natural to work with an external library without its proper react bindings could be a right way to bring it up. At this point I would still not cover anything about the `Trans` component. Only the `t` function. The question to tackle is, how can you make the `t` function available as a property of the component? The answer is `translate` HOC. I know that there are more than one possibility to do this, like the `I18n` component but at least for me the integration of a HOC felt more natural. My assumption is that most of the people will feel this way too. 
     * Now it's time to introduce the `Trans` component. To do this, I'd bring up the problem that it solves: Providing a nested React component that would require a too repetitive use of the `t` function, highlighting its advantage. However, still state that it's by no means required.
     * Last, connect the dots by making the developer think about a broader use of the `t` function. Why is better to use the Provider over importing the  `translate` HOC in each component? I know that this might be more related to the React documentation but still it adds up on value to `react-i18next`'s documentation. 

* Like I mentioned before, I'd stick throughout this 4 steps with just one variant. About the rest of the documentation, I'm still thinking about it, so far my ideas are: 
    1. Group the components by what they deliver, e.g. `t` function and `Trans` component are for resolving keys that hold the translations therefore they should be both in one
    2. Have a thing like a glossary to explain the different components, functions, etc.. and add the overview there. 

What do you think? Happy to discuss and hear your opinions.

**Comment by jamuhl** [comment]
@juanecabellob 
- for sure agree on the tutorial page you described...guess makes a lot of sense and would help getting started.
- grouping really might make sense - so it's more clear which component are directly related to translation and which are more a layer of additional support (hoc, provider)
- glossary would not hurt (would help finding stuff faster)

Will need some time to add this...but i like the ideas and will try getting that done in the next weeks.

**Comment by juanecabellob** [comment]
Sorry, I've been very busy. Haven't forgotten it, though. Maybe next week will free up a bit.

**Comment by jamuhl** [comment]
don't worry...no need to rush it.

**Comment by alvaromb** [comment]
I don't want to pollute this issue but I would like to highlight one important thing in @juanecabellob's excellent writeup about how the docs should explain what this library does.

In his last step, he describes the use of the Provider component, but right now in the docs one cannot exactly distinguish which are the differences about using Provider or `reactI18nextModule`, or even if they're the same.

Also, a full example of a configured app with `reactI18nextModule` is missing too. It seems that either you use it with Provider or with the HOC, but then again, it is stated as an alternative to the Provider.

**Comment by jamuhl** [comment]
most full examples uses it: https://github.com/i18next/react-i18next/blob/master/example/react_withHOC/src/i18n.js#L9

but you're right the documentation does not have a sample using the reactI18nextModule.

improvements to the docs are very welcome - if no one else does them i will do those as soon i got time...(but as you might guess - this is not my only OSS repo and beside job and family my time is limited too)

**Comment by alvaromb** [comment]
Sure, @jamuhl, I appreciate a lot your work here. Please let me integrate first the library into one of our projects and then I'll check if I can send some suggestions or improvements to the docs.

**Comment by jamuhl** [comment]
we just published v8.0.0 of react-i18next relevant here:

- translate hoc got renamed to withNamespaces (translate got marked deprecated)
- I18n render prop got renamed to NamespacesConsumer (I18n got marked deprecated)
- Interpolate got marked deprecated

hope the new namings make it more clear what the component really does.

**Comment by jamuhl** [comment]
@alvaromb @juanecabellob just made my attempt to improve the documentation:

https://react.i18next.com/guides/the-drawbacks-of-other-i18n-solutions
https://react.i18next.com/guides/quick-start
https://react.i18next.com/guides/step-by-step-guide

Feedback would be very welcome

**Comment by alvaromb** [comment]
Thanks @jamuhl, will check later today.

**Comment by juanecabellob** [comment]
I had a first glimpse and looks better! Will read it thoroughly later today, as well.

**Comment by juanecabellob** [comment]
I had already a first glimpse, it looks way better! I will read thoroughly later today, as well.

**Comment by alvaromb** [comment]
I think it's a step into the right direction, @jamuhl. My initial problems with the docs are most of all solved.

**Comment by jamuhl** [comment]
@alvaromb nice to hear....if you got any suggestions for improving...just do a PR...gladly would merge those.

closing this for now...

**Comment by csergiu** [comment]
The Trans component documentation is still incredibly unclear. How do you use it? The examples given are very weird: <1></1> and then <5></5>. What do those mean? I can't wrap my head around it. Also, why do you use the same text from the translation file inside the component? The entire point of translating it is to tie it to a key and then let the translator handle it.

**Comment by jamuhl** [comment]
https://react.i18next.com/latest/trans-component#how-to-get-the-correct-translation-string

**Comment by tenthirtyone** [comment]
**Suggestion**: Follow your own docs step-by-step and fix them before releasing them

You could do this yourself before publishing the docs but I'll do it and copy and paste every problem that I hit.Your own warnings are calling out your docs...

Literally following the step-by-step docs results in:

```
i18next: hasLoadedNamespace: i18next was not initialized
i18next.js:27 i18next::translator: key "blah" for namespace "translation" for languages "en" won't get resolved as namespace was not yet loaded This means something IS WRONG in your application setup. You access the t function before i18next.init / i18next.loadNamespace / i18next.changeLanguage was done. Wait for the callback or Promise to resolve before accessing it!!!
i18next::translator: missingKey undefined translation blah blah
i18next::backendConnector: loading namespace translation for language es failed failed parsing /locales/es/translation.json to json
i18next::backendConnector: loading namespace translation for language en failed failed parsing /locales/en/translation.json to json
```

If you are going to use other libraries in your example you are responsible for telling us how to use them. You don't get to link to the github repository and say "here you figure it out"

**Suggestion:** Literally put these two lines in your step-by-step: 

```
create a /public/locales/{lan}/translation.json file
put   { "title": "title" } in translation.json
```

But then you get this beauty:

```
i18next: hasLoadedNamespace: i18next was not initialized undefined
i18next.js:27 i18next::translator: key "blah" for namespace "translation" for languages "en" won't get resolved as namespace was not yet loaded This means something IS WRONG in your application setup. You access the t function before i18next.init / i18next.loadNamespace / i18next.changeLanguage was done. Wait for the callback or Promise to resolve before accessing it!!!
i18next::translator: missingKey undefined translation blah blah
i18next::backendConnector: loaded namespace translation for language en {title: "title"}
i18next::backendConnector: loaded namespace translation for language es {title: "title"}
i18next: languageChanged es
i18next: initialized {debug: true, initImmediate: true, ns: Array(1), defaultNS: Array(1), fallbackLng: Array(1), …}
```

It's broken. Please fix it and try it first, before pushing it live.

**Comment by jamuhl** [comment]
@tenthirtyone 

a) the console tells you step by step what is wrong.
b) https://react.i18next.com/latest/using-with-hooks#configure-i-18-next tells in the comment you have to look up more details for backend usage in that repo
c) it's a step by step guide not a full repository sample -> if you need that there are plenty of them in the repo (like said the console points you to the things going wrong - both documentation and console experience were improved a lot - sadly seems still be not what one expects from free software)
d) your last problem not following the step by step guide -> you did something wrong (not using Suspense, hook or HOC)
E) This is OSS - so I would suggest to tone down your language a little (We/I don't owe you anything - If something can be improved feel free to share a PR instead of blaming the quality of a free cake)
F) If you like to consume OSS and not contributing to it consider:

If you like this module don’t forget to star this repo. Make a tweet, share the word or have a look at our https://locize.com to support the devs of this project.

If you liked my support / work - I would enjoy a coffee sponsored using the “:purple_heart:Sponsor Button”.

There are many ways to help this project :pray:

**Comment by tenthirtyone** [comment]
a) The console tells me what is wrong, without any clue as to how to fix it and the point of a step-by-step guide is not to end up with a broken final step.

b) Then put the comments in the step-by-step guide you knuckle dragger

c) A step by step guide is meant to get you to something that works. Not something that is broken  that requires you to go lookup other guides.

d) I literally copied and pasted your guide. You likely have a broken install in some whatever is breaking for everyone else following this guide doesn't have the same global libs installed or something other junk. I suggest you take a test system or fresh system and actually test your code.

e) You don't owe me anything but when you push broken code and people waste time on it, you are responsible for your broken code and the waste of time.

f) Get off your cross. It's technically not OSS because it doesn't work. Until it works it's not code. If you want to wrap yourself in the OSS mantle you need to release working OSS.

I don't support repos or code that is unmaintained. It's over engineered config file. You've managed to over engineer and break the simple instructions for config file substituting values in text. You've broken copy and paste at the software level.

Do you understand the impact of 1 hour delay across 1 hour developers? It's broken. Unfuck it.

@jamuhl It's an over-engineered config script

**Comment by juanecabellob** [comment]
@tenthirtyone if it's wasting your time, do it yourself. stop harassing people. if it's broken, then look somewhere else. if you want to make it better, contribute. if you wanna get mad, be mad at yourself for allowing yourself to be mad at software or look for another profession.

**Comment by juanecabellob** [comment]
@tenthirtyone be constructive, not destructive. you're not getting anywhere and just making things worse talking like that. yes, there's a delay. but then what? is it the end of the world? no.

**Comment by tenthirtyone** [comment]
@juanecabellob harassment is aggressive pressure of intimidation - it's what you're doing right now. Get a life

I'm getting my time back from this broken, shoddy repo.

**Comment by tenthirtyone** [comment]
Again, pointing out, you over engineered and broke string substitution and then you couldn't even write documentation because you broke and over complicated it that bad.

**Comment by juanecabellob** [comment]
@tenthirtyone good luck buddy. i don't think you'll get that time back and just get yourself ignored.

**Comment by tenthirtyone** [comment]
Can you imagine the distributed dev hours spent wasted trying to figure out your broken 'OSS' and documentation before giving up? 

How about spending half the time asking for stars fixing the problem? This is almost 2 years old

**Comment by tenthirtyone** [comment]
@juanecabellob I don't think or care about you

**Comment by juanecabellob** [comment]
@tenthirtyone give up and get another job. if you have to spend so much time figuring out a bug then software programming clearly not yours :)

**Comment by tenthirtyone** [comment]
I would tell you to RTFM but it's broken

**Comment by jamuhl** [comment]
@juanecabellob thank you for stepping in and sorry for your wasted time talking to @tenthirtyone 

@tenthirtyone Won't argue with you - you're absolutely right...I'm ashamed of putting this overengineered  i18next on github...this piece of overcomplicated config file thing thousands of projects rely on making a shit ton of money (talking about big companies, companies in the banking, insurance and what ever sector). But you're right...I failed...I hope you teach me better and let me learn from your wisdom...looking forward to your i18n approach...

**Comment by tenthirtyone** [comment]
@jamuhl oh yeah, I bet if you just stopped today the line out the door would be amazing.
How many powers of 10 is 'shit ton' because my guess is engineers are told they have to do i18n and then get stuck with this garbage wasting most their time before giving up for going elsewhere.

You have under 5,000 weekly downloads compared to your parent repo that gets 650,000. That's the biggest indicator you fucked up unless you want to argue no one uses react

**Comment by adrai** [comment]
Please let's stay on a factual level.
@tenthirtyone What exactly is the concrete problem? How could we improve it? Would you be willing to help? Can you share a https://codesandbox.io example?
Please help us to help you.

**Comment by tenthirtyone** [comment]
Literally read my post

**Comment by adrai** [comment]
Done.
This (https://github.com/adrai/react-i18next-step-by-step) is my result after following this page: https://react.i18next.com/latest/using-with-hooks
I have no strange console statements. 🤷‍♂️

**Comment by naemtl** [comment]
Just wanted to chime in here and say that the Trans Component docs are still very unclear and hard to understand. I've wasted over 6 hours trying different ways to make the results I want come through via Trans component usage, but the docs make it very hard and way more time consuming than it should be. Please fix the docs!

**Comment by adrai** [comment]
If you have any idea to make it better, feel free to contribute.

**Comment by naemtl** [comment]
@adrai I would recommend you create a fool proof, step by step process as to how the most recent iteration of the Trans component works. 

For example, there's no mention of the **ns** attribute in [this](https://react.i18next.com/latest/trans-component) doc at all, I had to bust my head trying to figure out that I can destructure an array of elements in the **components** attribute which I could then access via indices in my translation JSON file. So like: 

`<Trans
      components={[
        <strong />
      ]}
      count={myCountVariable}
      i18nKey="keyHere.subKeyHere"
      ns="nameOfJsonFileHere"
      parent="typeOfElementToRenderTextIntoHere IE: div/span/p/etc"
  />` 
There's no mention of the way I'm using the **components**, **ns**, or **parent** attributes in the docs in the way that I am here above. 

Furthermore:

`
...
"subKeyHere": "<0>{{count}} block</0> of tofu is in my fridge right now."
`

The way you describe using the components attribute within the JSON files is confusing as well. You can just make it more clear that the first element you pass into the attribute is <0>, and so on. So in my example, <0>{{count}} block</0> in my JSON formatted string is a placeholder for <strong>{{count}} block</strong> and will make my "{{count}} block" text bold."

Come on guys, it's been almost 3 years that people are telling you about this and nothing has changed.

**Comment by adrai** [comment]
@naemtl 

> there's no mention of the **ns** attribute in [this](https://react.i18next.com/latest/trans-component) doc at all

Maybe I'm looking at the wrong place, but it's all here: https://react.i18next.com/latest/trans-component#trans-props
![image](https://user-images.githubusercontent.com/1086194/100791323-6350d280-3419-11eb-89ff-f123ef6f4ad4.png)

> The way you describe using the components attribute within the JSON files is confusing as well. You can just make it more clear that the first element you pass into the attribute is <0>, and so on.

I don't get what you mean. Are you talking about this? https://react.i18next.com/latest/trans-component#how-to-get-the-correct-translation-string

Maybe the docs have to be written by someone who doesn't know react-i18next (and i18next) yet. 🤷‍♂️
If you like to, please feel free to click on the "Edit on GitHub" button and propose an improvement.
Any help is welcome.

**Comment by naemtl** [comment]
The table does help, but I feel like it would be even more helpful to include more examples of each prop being used.

**Comment by naemtl** [comment]
Also, maybe include the table at the top of the document instead of the very bottom!

**Comment by jamuhl** [comment]
Closing this thread for now...as it already started with too much heat.

@everyone feel free to contribute improvements to the docs to make the lives of the ones that follow you easier.
